### ISSUE TITLE
`ushort` `ContractState.UpdateCounter` overflow

### ISSUE THREAD

**Issue by vang1ong7ang** [neo-project/neo#2668]
AN EPIC ACHIEVEMENT IS UNLOCKED:

by keeping updating the smart contract until its`UpdateCounter` jumps back to `0`

although it is designed to be increased only

it is proven that `UpdateCounter` is not secure any more with the hundreds of GAS cost.

i suggest to use something longer than `int` to replace `ushort`.

**Comment by vang1ong7ang** [comment]
the script hash of the overflowed contract is: `0x9f89a0f5de46a085c99e6ef54a977bb37b2099b1` on neo3 testnet

the code can be something like

```
using Neo.SmartContract.Framework;
using Neo.SmartContract.Framework.Native;
using Neo.SmartContract.Framework.Services;

public class NotificationBomb : SmartContract
{
    public static void Update(ByteString nef, string manifest)
    {
        ContractManagement.Update(nef, manifest, null);
        var count = ContractManagement.GetContract(Runtime.ExecutingScriptHash).UpdateCounter;
        for (int i = count; i < 65536; i++)
        {
            ContractManagement.Update(nef, null, null);
        }
    }
}
```

**Comment by roman-khimov** [comment]
Fun as it is I doubt it has any practical consequences. Legitimate token TOK contract's owner won't be updating it in a loop like this (if someone else is doing it to TOK then the control is lost already) and he's not very likely to reach this limit even after years of operation. Specially-crafted contract will overflow, but who's gonna use it?

**Comment by vang1ong7ang** [comment]
i think there would be no critical risk, however, it can cheat users or explorers.

i.e. onegate browser has a verified source code section and can be cheated through this way: <https://testnet.explorer.onegate.space/contractinfo/0x9f89a0f5de46a085c99e6ef54a977bb37b2099b1>
(however it is crashed before the cheat kkkk)

to fix it, the cheapest way can be: change the following line

https://github.com/neo-project/neo/blob/1e449185c0d05e9824c257ab4404f380106d15b2/src/neo/SmartContract/Native/ContractManagement.cs#L239

to

```cs
checked { contract.UpdateCounter++; }
```

**Comment by shargon** [comment]
> by keeping updating the smart contract until itsUpdateCounter jumps back to 0

You can destroy and create the contract again to to get UpdateCounter=0

**Comment by vang1ong7ang** [comment]
> You can destroy and create the contract again to get UpdateCounter=0

👍 haha it is absolutely a better way, but it clears the contract storage and using a new contract id.

**Comment by shargon** [comment]
> using a new contract id.

This is transparent to the user, the address could be the same. I think that overflow `UpdateCounter` is not problematic, is just a counter, updating the contract the owner could produce more damage to the user.

**Comment by devhawk** [comment]
Being able to reset the `UpdateCounter` to zero, either by overflow or destroy + recreate, seems to defeat the purpose of having `UpdateCounter`. I get asked regularly (including in my devlog presentation today) about the security risk of supporting updatable contracts. My usual response is that the update counter can be checked to ensure the contract is the expected version. But if that value can be `UpdateCounter` manipulated, an attacker could disguise their contract update and trick users into invoking a malicious contract.

Maybe we should be exposing the NEF checksum or a SHA of the NEF script contents that could be used for optionally validating script contents for developers that wish to guard against this kind of attack?

cc @erikzhang @Liaojinghui

**Comment by Liaojinghui** [comment]
Well, indeed it is an overflow, but overflow does not necessarily mean a vulnerability.  I dont see what attackers can do with this counter, what is the difference between 0 and 10000,,,,,,,especially attacker has to be the owner of the contract.

But, it we allow users to verify the version of the contract from transaction, it will be a different story.

**Comment by devhawk** [comment]
> I would vote to update `ushort` to `uint`~~~~~

Would that solve the delete / recreate issue?

**Comment by Liaojinghui** [comment]
> Would that solve the delete / recreate issue?

what you mean `destroy + recreate`? are you saying redeploy? wont  operation like `recreate` with new contract code generates another contract address? different contract address means different contracts~~~

**Comment by Liaojinghui** [comment]
> The contract address is derived from NEF checksum + contract manifest name + account that deployed the contract. So it is possible to `destroy` a contract and then redeploy it and get the same contract address.

This is impossible but with negligible possibility........Hash function is supposed to be collision resistance,,,, we should not assume anyone could find two messages generate the same hash value. Otherwise, the whole Blockchain system breaks.

**Comment by Liaojinghui** [comment]
> Is there a way to ensure a destroyed contract address is not reused?

Actually we can, just maintain an address list of destroyed contracts or would solve this problem... the core issue here is that we need a mechanism to update existing native contract.

**Comment by erikzhang** [comment]
> update contract same number of times from step 2, but this time to a malicious version
> You now have a malicious contract deployed in a manner that cannot be detected.

You can detect it by checking the `Update` event of `ContractManagement`.

https://github.com/neo-project/neo/blob/41055be286fee8ee49bd64e9cf3c47031489264c/src/neo/SmartContract/Native/ContractManagement.cs#L99

**Comment by Liaojinghui** [comment]
But it is not a runtime detection,  a user who does not run a full node could not know this event.

**Comment by erikzhang** [comment]
The wallet should warn the users in this case.

**Comment by Liaojinghui** [comment]
then its more like a feature and option of wallet, instead of neo. From a security view, neo is vulnerable and relies wallet to patch it.

**Comment by devhawk** [comment]
> You can detect it by checking the Update event of ContractManagement.

I don't understand how this would be any different than looking at the `UpdateCounter` value. 

You *could* detect the [exploit I described](https://github.com/neo-project/neo/issues/2668#issuecomment-1099229878) by looking for a `ContractManagement.Deploy` event for a contract hash that had previously raised a  `ContractManagement.Destroy` event.

I agree with @Liaojinghui that this isn't something that should be delegated to wallets to enforce. Beyond the basic issue of delegating security enforcement to external systems, I don't see how a light wallet could access the entire ContractManagement event history to search for the `Destroy` + `Deploy` pattern.

Even if every light wallet could be trusted to enforce this correctly, there's still the question of contract -> contract invocation. As far as I know, there is no way to access ContractManagement event history from inside a contract. Even if you could, I suspect developers wouldn't be willing to pay the GAS cost to checking every ContractManagement event to make sure the contract they're about to invoke hasn't got a suspicious `Destroy` + `Deploy` pattern.

**Comment by devhawk** [comment]
Using OneGate Explorer, I was able to identify two contracts on MainNet that have been destroyed: 
* [FTW Token Test](https://explorer.onegate.space/contractinfo/0xdfb59623ef587b78ea1ec8a2d469171526b52b90) {[Destroy TX](https://explorer.onegate.space/transactionInfo/0x25587df8c1b996c3a0a008404cdc068e195d68fae6293d76638daaefad08ca4a)}
* [SkyContract](https://explorer.onegate.space/contractinfo/0xaee4031951a0a4934d81b15fc4e314d0a52822a5) {[Destroy TX](https://explorer.onegate.space/transactionInfo/0x25587df8c1b996c3a0a008404cdc068e195d68fae6293d76638daaefad08ca4a)}

Obviously "DevHawk eyeballs it" isn't a good way to determine the full list of destroyed contracts (plus, testnet has 10x as many ContractManagement events, and I'm not looking thru that many pages of event manually). But this does suggest that the number of destroyed contracts is small. Can we build a mechanism to populate the destroyed contract history w/o requiring a node resync?

**Comment by devhawk** [comment]
Fixed by #2697, we should close this issue
